client/landing/gutenboarding/stores/onboard/reducer.ts

```diff
@@ -1,62 +1,48 @@
 /**
  * External dependencies
  */
-import { Reducer } from 'redux';
 import { combineReducers } from '@wordpress/data';
 
 /**
  * Internal dependencies
  */
-import { ActionType, SiteVertical } from './types';
-import * as Actions from './actions';
+import { ActionType, OnboardAction, SiteVertical } from './types';
 
-const domain: Reducer<
-	import('@automattic/data-stores').DomainSuggestions.DomainSuggestion | undefined,
-	ReturnType< typeof Actions[ 'setDomain' ] >
-> = ( state = undefined, action ) => {
+type DomainSuggestion = import('@automattic/data-stores').DomainSuggestions.DomainSuggestion;
+type Template = import('@automattic/data-stores').VerticalsTemplates.Template;
+
+function domain( state: DomainSuggestion | undefined = undefined, action: OnboardAction ) {
```

Comment:
Setting the `state`'s default value to `undefined` is redundant.

Also, using the `Reducer<>` type has some benefits, as it adds `# | undefined` automatically to the state type:
```js
const domain: Reducer< DomainSuggestion, OnboardAction > = ( state, action ) => {
  ...
}
```

I continue to be very confused about the `OnboardAction` type. I believe that the action type shouldn't be constrained in any way beyond requiring an object with `type` field. And that's what the `import { Action } from 'redux'` type specifies.

Redux will call the reducer with an init action like this one:
```js
{
  type: "@@redux/INITj.3.e.w.g"
}
```
This action doesn't pass typecheck for `OnboardingAction`, and yet the reducer is guaranteed to be called with it and must react to it correctly.

On the other hand, I see exactly this kind of constraining the action type in the official docs, too:
https://redux.js.org/recipes/usage-with-typescript/#type-checking-reducers
Is it a bug, or am I just clueless about something?

Reply:
The `Reducer` type is very strange and I've wondered about this myself. We _claim_ that a `Reducer` handles a specific action type, but as you say it must handle _all_ actions.

I'd speculate it's a tradeoff the authors decided to make because by _convention_ —you _can't_ use Redux if you don't understand the `if ( action.type === 'MY_ACTION' ) { return 'new state' } else { return state }` pattern—, reducers gracefully handle all action types.

If we expect this by convention and lie a little bit to the type system, we get better information in our reducer functions. A correctly typed reducer function that wants to deal with a specific action but has to deal with actions of `{ type: any }` ends up being painful to work with. Consider that we can't even get a discriminated union because the basic action type `{ type: any }` also satisfies a fully typed `{ type: 'MY_ACTION', foo: string, bar: number, baz: boolean }`, so in our `if ( action.type === 'MY_ACTION' ) { … }` branch, we still don't know whether which type we have!

I think the following illustrates it:

```ts
import { Reducer } from 'redux'

interface ReducerCorrect<S, A> {
    // Note the `A | A_` here, our type effectively opens up to `{ type: any }` 😞
    <A_ extends { type: any; }>(state: S | undefined, action: A | A_): S;
}


type Action = { type: 'foo', bar: { baz: { quux: number } } }

// This is unsafe, but the experience is better
const reducerRedux: Reducer<number, Action> = (state = 0, action) => {
     // Whoa, that's going to error at runtime for sure!
    // Of course, anyone experienced with Redux won't write this…
    return action.bar.baz.quux;
}

// This is safe, but is a pain
const reducerCorrect: ReducerCorrect<number, Action> = (state = 0, action) => {
    if (action.type === 'MY_ACTION') {
        return action.bar.baz.quux; // We can't know this -- type error! 😭
    } 
    return state;
}
```

Reply:
My thoughts about typing tradeoffs echo @sirreal's. The global nature of Redux actions almost makes them untypable. Elm's pattern of using nested actions is a way of making all the types perfect, but global events are pretty handy :)

> Setting the state's default value to undefined is redundant.

Good point, I'll clean that up while I'm here.

> using the Reducer<> type has some benefits, as it adds # | undefined automatically to the state type

I actually like that not all state types are `# | undefined`. Yes the onboard store has a lot of optional pieces of state so it'd be a convenience here. But the `domainSuggestion` state is never `undefined` and I feel like a lot of reducers work this way e.g. a lot of reducers are just a single value type and have a default.

Reply:
🤔After considering this a bit more and seeing some other context (p1579980641021900-slack-typescript), I'd change my speculative stance 🙂 

The second type argument to `Reducer` is fine if left to the default. But, if it becomes the union of the type of _all actions in your application_, it should give excellent information and type safety.

Reply:
> Setting the `state`'s default value to `undefined` is redundant.

I've pushed a fix for this.

> The second type argument to `Reducer` is fine if left to the default. But, if it becomes the union of the type of all actions in your application, it should give excellent information and type safety.
 
@sirreal Just to clarify, does that mean you're happy with the changes here since it allows reducers to accept any of the actions for a given store? I thought that was your original (speculative) stance.

Reply:
> does that mean you're happy with the changes […]

I've been considering these changes and playing with the types but haven't found the time to leave a full review. I'd prefer to leave it for a while.

Reply:
It seems that the TypeScript types for Redux reducers are buggy. When I created a toy example of a reducer that accepts only a constrained set of actions, it indeed fails to compile valid dispatches of unknown actions:
```js
function reducer( state, action: HandledActionTypes ) {
  ...
}
reducer( undefined, { type: '@@INIT' } );
createStore( reducer ).dispatch( { type: '@@INIT' } );
```
Both these statements fail to compile and compiler reports an error:
```
Type '"@@INIT"' is not assignable to type '"ACTION_A" | "ACTION_B"'
```
The Redux repo already has a similar issue: https://github.com/reduxjs/redux/issues/3580
It describes another funny consequence of the constrained action type: one can write a reducer that fails to handle unknown actions:
```js
function reducer( state: MyState = initState, action: MyActions ): MyState {
  switch ( action.type ) {
    case MY_ACTION_A:
      return { ...state, ... };
    case MY_ACTION_B:
      return { ...state, ... };
  }
}
```
The switch statement exhausts all possible values of `action.type` for the `MyActions` type, so TypeScript doesn't mind that it returns `undefined` for all other `action.type`s. It thinks it can never happen 🙂 But such a reducer breaks the Redux contract, and also fails to always return a value of type `MyState`.

I added a comment that describes our case: https://github.com/reduxjs/redux/issues/3580#issuecomment-579239622

Reply:
Thanks for pointing out that Redux issue thread @jsnajdr, we're not the only one's grappling with how Redux and TS should work together.

I took a look at the [solution Simplenote is using](https://github.com/Automattic/simplenote-electron/blob/develop/CONTRIBUTING.md#redux-and-typescript), and their typings suffer from the same issue i.e. it's possible to write a reducer with no type errors that doesn't deal with unknown action types.

This isn't ideal, but given it's something community hasn't figured out yet (the approach recommended by the docs has the issue) is this something where we just wait and see where Redux goes with this. This PR improves things a little at least by widening the action type.